### ISSUE TITLE
PostgreSQL support

### ISSUE THREAD

**Issue by sinzone** [Mashape/kong#331]
We should support other datastores than Cassandra which is primarily aimed at users who want to benefit from its distribution capabilities.

Support for PostgreSQL seems the best candidate for the second datastore, due to its popularity, availability on Amazon RDS and reliability. It is also a very good fit for Kong due to its relational nature.

**Comment by sinzone** [comment]
related #183

**Comment by thorstenz** [comment]
+1 for PostgreSQL support.
very good experience with the jsonb datatype.

**Comment by rainavishal1** [comment]
+1

**Comment by bhameyie** [comment]
+1

**Comment by MattCollinge** [comment]
+1

**Comment by skippy** [comment]
+1

**Comment by wsimmonds** [comment]
+1

**Comment by obsidienne** [comment]
+1

**Comment by lucamaraschi** [comment]
:+1:  :+1:

**Comment by sent-hil** [comment]
:+1:

**Comment by matti** [comment]
:+1:

**Comment by ahmadnassri** [comment]
:+1:

**Comment by vivekfantain** [comment]
+1

**Comment by owahlen** [comment]
+1

**Comment by coreybutler** [comment]
+1

**Comment by ngbinh** [comment]
:+1:

**Comment by bigpg** [comment]
+1

**Comment by udlei** [comment]
+1

**Comment by avtar** [comment]
:+1:

**Comment by vwal** [comment]
+1

**Comment by bkleef** [comment]
+1

**Comment by edurdias** [comment]
+1

**Comment by devton** [comment]
+1

**Comment by mgooty** [comment]
+1

**Comment by soulware** [comment]
+1 - would be nice to be able to simply use an RDS instance on AWS for example.

**Comment by schoenobates** [comment]
+1

**Comment by harlow** [comment]
+1

**Comment by dan-turner** [comment]
+1!!

**Comment by shinzui** [comment]
Is this on the roadmap?

**Comment by sinzone** [comment]
@shinzui it is.

Plus there is @jerizm (the hero) working on it 
https://github.com/Mashape/kong/compare/master...jerizm:feature/postgres

**Comment by jerizm** [comment]
Thanks @sinzone. Not a hero quite yet. It is quite long and complicated. Good news is it works (mostly) but there are still a lot of things to be done!

**Comment by jbgo** [comment]
:+1:

**Comment by texasviking** [comment]
:+1:

**Comment by karanshah89** [comment]
+1

**Comment by t1tcrucible** [comment]
+1

**Comment by ahmadnassri** [comment]
we're locking this conversation, this is coming in [`0.6.0`](https://github.com/mashape/kong/wiki)

**Comment by e-durante** [comment]
+1

**Comment by minddust** [comment]
+1

**Comment by johnnyt** [comment]
+1

**Comment by diwadm** [comment]
+1

**Comment by sofiaguyang** [comment]
+1

**Comment by leandrosilvaferreira** [comment]
+1

**Comment by vanettem** [comment]
+1

**Comment by RussellLuo** [comment]
+1

**Comment by JnMik** [comment]
+1

**Comment by kenwarner** [comment]
anything in particular that's left to do in feature/postgres?

**Comment by sinzone** [comment]
@kenwarner invited you to our new Postgres chat. 

also: https://hackpad.com/Cassandra-to-Postgres-EbuFvpPMj5u

**Comment by gregory** [comment]
:+1:

**Comment by z0mbix** [comment]
+1

**Comment by markkennedy** [comment]
+1

**Comment by mgvarley** [comment]
+1

**Comment by rynam0** [comment]
+1.  MySQL would also be fantastic

**Comment by AlexMaskovyak** [comment]
+1

**Comment by Leventhan** [comment]
+1

**Comment by jet10000** [comment]
+1

**Comment by adamlwatson** [comment]
+1

**Comment by matttylr** [comment]
+1

**Comment by johnboy14** [comment]
+1000

**Comment by eric1iu** [comment]
+1

**Comment by calebmer** [comment]
+1

**Comment by khaight** [comment]
:+1:

**Comment by gudubeth** [comment]
+1

**Comment by cangus** [comment]
+1

**Comment by rhussmann** [comment]
+1

**Comment by lobsterdore** [comment]
:+1:

**Comment by wanghaisheng** [comment]
+1

**Comment by nazarhussain** [comment]
+1

**Comment by cwanjau** [comment]
+1

**Comment by ababkov** [comment]
+1

**Comment by ialoft** [comment]
+1 when we start ?

**Comment by micyoung70** [comment]
+1

**Comment by narrativecraig** [comment]
+1

**Comment by sfoulkes** [comment]
+1

**Comment by mtrahan** [comment]
+1

**Comment by poori** [comment]
+1

**Comment by akurkin** [comment]
+1

**Comment by msmathers** [comment]
+1

**Comment by marianudo** [comment]
+1

**Comment by talldave** [comment]
+1

**Comment by danlopez00** [comment]
+1

Also a Redis and or ElasticSearch backend would be amazing. Please look at what AWS RDS has, so that a team does not have to worry about scaling the backend

**Comment by andyfleming** [comment]
+1 for supporting RDS instance on AWS. Aurora would be great too.

**Comment by abtreece** [comment]
+1 PostgreSQL (RDS)

**Comment by siddharthkchatterjee** [comment]
+1

**Comment by RolfKoenders** [comment]
+1

**Comment by thibaultCha** [comment]
I have been working on integrating Postgres for a few days now, and I would like to know which version of Postgres people use, and if they are locked on that specific version or not. Since RDS supports 9.4, I take it as a good metric but I would still like to know if people use an older version, and if they're not willing to upgrade.

Feedback is appreciated.

**Comment by owahlen** [comment]
Hello Thibault,
it is very good news that work on postgres has started!
In our team we use a recent version of postgres from AWS.
Currently we are not bound to a specific version or feature.

Best regards,
Oliver



Am 28.01.2016 um 04:52 schrieb Thibault Charbonnier <
notifications@github.com>:

I have been working on integrating Postgres for a few days now, and I would
like to know which version of Postgres people use, and if they are locked
on that specific version or not. Since RDS supports 9.4, I take it as a
good metric but I would still like to know if people use an older version,
and if they're not willing to upgrade.

Feedback is appreciated.

—
Reply to this email directly or view it on GitHub
<https://github.com/Mashape/kong/issues/331#issuecomment-175952657>.

**Comment by thorstenz** [comment]
Hello Thibault,

great news!

Postgres version 9.4 is in use here.
Not bound to a specific version: everything >= 9.4 would be super.

Best

Thorsten

**Comment by khaight** [comment]
>= 9.4. This is HUGE!

**Comment by abtreece** [comment]
We are using 9.2+. Would use RDS for our Kong installation so that is a good baseline!

**Comment by supermario** [comment]
We are using 9.4. :smiley:

**Comment by siddharthkchatterjee** [comment]
We are using 9.4 as well, on RDS.

**Comment by andyfleming** [comment]
We aren't using Postgres for anything now. We're happy if it runs on RDS.

**Comment by RolfKoenders** [comment]
We are using 9.4 as well, on RDS.

**Comment by pklingem** [comment]
9.4 is great for us.

**Comment by C4** [comment]
Like others, Not using Postgres right now. But something RDS backed would be awesome.

**Comment by lindo-jmm** [comment]
Not on AWS (Azure), but I have 9.4 running on a server. Love it!

**Comment by nhuray** [comment]
+1

**Comment by kenwarner** [comment]
9.4 on RDS is our use case as well.

**Comment by dempe** [comment]
:+1: Any version would be fine for us.

**Comment by protetore** [comment]
We're using 9.3 on RDS.

**Comment by stramel** [comment]
+1

**Comment by derrickburns** [comment]
:+1:

**Comment by shaikhfaisal** [comment]
+1 RDS support would be awesome

**Comment by mjhlobdell** [comment]
+1

**Comment by lsmith77** [comment]
@thibaultCha is there an ETA? also is there already a PR to try out your current state?

**Comment by johannes-pt** [comment]
+1

**Comment by jkwong** [comment]
+1

**Comment by sathyavijayan** [comment]
+1

**Comment by eseres** [comment]
+1

**Comment by bramswenson** [comment]
:+1:

**Comment by gianlucatomasino** [comment]
+1

**Comment by nickmcintyre** [comment]
+1

**Comment by bad79s** [comment]
+1

**Comment by jaceq** [comment]
Is there maybe any update on this topic?

**Comment by BrianHutchison** [comment]
+1 for Amazon DynamoDB.

**Comment by thibaultCha** [comment]
Hey everyone,

The implementation is almost complete, the branch now lives in this repository (See [feature/postgres](https://github.com/Mashape/kong/tree/feature/postgres)). As planned, Postgres 9.4 is supported so Amazon RDS should be supported as well, I briefly played with it but it needs to be tested more in-depth once the implementation is complete. Btw, testers are welcome here!

###### What is missing:
- Full support for OAuth2 and clustering. Right now those implementations relied on Cassandra rows with TTLs, so right now with Postgres, things don't expire as they would do with Cassandra. This should soon be resolved. This block production usage but not testing of the support.
- Support for SSL connections between Kong and Postgres. As of now, a Pull Request is pending at leafo/pgmoon#24 with support for it. If it gets approved, that should be resolved too.
- A lot of documentation updates!
- Eventually, support for MD5 or Trust authentication (maybe not in the initial release)

###### How to test it (if you want to try it out or just a glimpse of how to use it):
- Get your hands on a 9.4+ server
- Create a database, (maybe a user too?), let's say `kong`
- Use the `feature/postgres` branch of Kong (you must have a source install for this)
- Update your Kong configuration:

```yaml
# be careful about your YAML formatting
database: postgres
postgres:
  host: "127.0.0.1"
  port: 5432
  user: kong
  database: kong
```

As usual, migrations should run from `kong start`, but as a reminder and just in case, here are some tips:

Reset the database with (careful, you'll lose all data):
```
$ kong migrations reset --config kong.yml
```

Run the migrations manually with:
```
$ kong migrations up --config kong.yml
```

If needed, list your migrations for debug purposes with:
```
$ kong migrations list --config kong.yml
```

If you test it, feel free to report any issues (please not here, but in a new issue in this repo) :)

Thank you all for your patience!

**Comment by geovanisouza92** [comment]
+1

**Comment by thefosk** [comment]
This feature will land in v0.8.

**Comment by thefosk** [comment]
You can play with v0.8.0 with Postgres support at https://github.com/Mashape/kong/releases/tag/0.8.0 - we will update Docker and the website soon.

**Comment by thefosk** [comment]
Kong 0.8.0 announced: http://blog.mashape.com/announcing-kong-0-8-0-with-postgresql-support/

Enjoy :)
